File: gank/src/main/java/com/github/wally/gank/mvp/view/GankMeiZiDetailFragment.java
package com.github.wally.gank.mvp.view;

import android.net.Uri;
import android.os.Environment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AccelerateInterpolator;

import com.github.wally.base.util.ImageDisplayUtil;
import com.github.wally.base.util.StatusBarUtil;
import com.github.wally.gank.R;
import com.github.wally.base.base.BaseMvpFragment;
import com.github.wally.base.base.BasePresenter;
import com.github.wally.gank.bean.gank.DisplayMeiZiImageBean;
import com.github.wally.gank.mvp.contract.GankMeiZiDetailContract;
import com.github.wally.gank.mvp.presenter.GankMeiZiDetailPresenter;
import com.github.wally.base.util.ImageDownloadUtil;
import com.github.wally.base.util.ToolBarHelper;
import com.github.wally.gank.widget.RotateCircleProgressBar;
import com.gyf.barlibrary.ImmersionBar;
import com.haoge.easyandroid.easy.EasyToast;

import java.io.File;

import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import uk.co.senab.photoview.PhotoView;
import uk.co.senab.photoview.PhotoViewAttacher;

/**
 * Package: com.github.wally.gank.ui.fragment
 * FileName: GankMeiZiDetailFragment
 * Date: on 2018/6/16  下午6:42
 * Auther: zihe
 * Descirbe:
 * Email: hezihao@linghit.com
 */
public class GankMeiZiDetailFragment extends BaseMvpFragment<GankMeiZiDetailContract.Presenter, GankMeiZiDetailContract.View> implements GankMeiZiDetailContract.View {
    private PhotoView mImageView;
    private Toolbar mToolbar;
    private RotateCircleProgressBar mLoadingView;

    private boolean isHideToolBar = false;
    private DisplayMeiZiImageBean mImageBean;

    @Override
    public void onDestroy() {
        StatusBarUtil.showStatusBar(getActivity());
        super.onDestroy();
    }

    @Override
    public int onLayoutId() {
        return R.layout.fragment_gank_meizi_detail;
    }

    @Override
    public void onFindView(View rootView) {
        super.onFindView(rootView);
        mImageView = findView(R.id.image_iv);
        mLoadingView = findView(R.id.loading_iv);
    }

    @Override
    public void onBindViewContent() {
        super.onBindViewContent();
        ToolBarHelper toolBarHelper = ToolBarHelper.newBuilder(getView(), R.id.tool_bar, new ToolBarHelper.ConfigCallbackAdapter() {
            @Override
            public void onConfigBefore(Toolbar toolbar) {
                ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);
            }
        })
                .withNavigationIconDrawable(R.drawable.ic_action_back)
                .withNavigationIconOnClick(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        getActivity().onBackPressed();
                    }
                })
                .build();
        mToolbar = toolBarHelper.getToolbar();
        ImmersionBar.with(this).titleBar(mToolbar);
        //长按点击事件
        mImageView.setOnPhotoTapListener(new PhotoViewAttacher.OnPhotoTapListener() {
            @Override
            public void onPhotoTap(View view, float x, float y) {
                toggleToolBar();
            }

            @Override
            public void onOutsidePhotoTap() {
            }
        });
    }

    private void toggleToolBar() {
        //隐藏
        if (!isHideToolBar) {
            mToolbar.animate()
                    .translationY(-mToolbar.getHeight())
                    .setInterpolator(new AccelerateInterpolator())
                    .start();
            StatusBarUtil.hideStatusBar(getActivity());
        } else {
            //显示
            mToolbar.animate()
                    .translationY(0)
                    .setInterpolator(new AccelerateInterpolator())
                    .start();
            StatusBarUtil.showStatusBar(getActivity());
        }
        isHideToolBar = !isHideToolBar;
    }

    @Override
    protected boolean setupSwipeBackEnable() {
        return true;
    }

    @Override
    protected GankMeiZiDetailContract.Presenter onCreatePresenter() {
        return new GankMeiZiDetailPresenter();
    }

    @Override
    public void onLayoutAfter() {
        super.onLayoutAfter();
        getPresenter().getMeiZiDetail(getArguments());
    }

    @Override
    public void showMeiZiDetail(DisplayMeiZiImageBean bean) {
        this.mImageBean = bean;
        mLoadingView.setVisibility(View.VISIBLE);
        ImageDisplayUtil.display(getActivity(), bean.getUrl(), mImageView
                , new ImageDisplayUtil.DisplayCallbackAdapter() {
                    @Override
                    public void onReady() {
                        super.onReady();
                        mLoadingView.setVisibility(View.GONE);
                    }
                });
    }

    @Override
    public void showError(Throwable throwable) {

    }

    /**
     * 保存图片到相册
     */
    private void saveImageToGallery() {
        if (mImageBean != null) {
            Disposable disposable = ImageDownloadUtil
                    .saveImageAndGetPathObservable(getActivity(), mImageBean.getUrl(), mImageBean.getCreatedAt())
                    .doOnSubscribe(new Consumer<Disposable>() {
                        @Override
                        public void accept(Disposable disposable) throws Exception {
                            showLoading("保存中，请稍后...");
                        }
                    })
                    .doOnTerminate(new Action() {
                        @Override
                        public void run() throws Exception {
                            hideLoading();
                        }
                    })
                    .subscribe(new Consumer<Uri>() {
                        @Override
                        public void accept(Uri uri) throws Exception {
                            File appDir = new File(Environment.getExternalStorageDirectory(), "Meizhi");
                            String msg = String.format(getString(R.string.picture_has_save_to),
                                    appDir.getAbsolutePath());
                            EasyToast.Companion.newBuilder().build().show(msg);
                        }
                    }, new Consumer<Throwable>() {
                        @Override
                        public void accept(Throwable throwable) throws Exception {
                            EasyToast.Companion.newBuilder().build().show("再试试...");
                        }
                    });
            ((BasePresenter)getPresenter()).addSubscription(disposable);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_detail, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_share:
                return true;
            case R.id.action_save:
                saveImageToGallery();
                return true;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: gank/src/main/java/com/github/wally/gank/mvp/view/GankMeiZiDetailFragment.kt
package com.github.wally.gank.mvp.view

import android.os.Environment
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.animation.AccelerateInterpolator
import com.github.chrisbanes.photoview.PhotoView
import com.github.wally.base.base.BaseMvpFragment
import com.github.wally.base.base.BasePresenter
import com.github.wally.base.util.ImageDisplayUtil
import com.github.wally.base.util.ImageDownloadUtil
import com.github.wally.base.util.ToolBarHelper
import com.github.wally.gank.R
import com.github.wally.gank.bean.gank.DisplayMeiZiImageBean
import com.github.wally.gank.ext.findView
import com.github.wally.gank.ext.hideStatusBar
import com.github.wally.gank.ext.showStatusBar
import com.github.wally.gank.mvp.contract.GankMeiZiDetailContract
import com.github.wally.gank.mvp.presenter.GankMeiZiDetailPresenter
import com.github.wally.gank.widget.RotateCircleProgressBar
import com.gyf.barlibrary.ImmersionBar
import com.haoge.easyandroid.easy.EasyToast
import java.io.File

/**
 * Package: com.github.wally.gank.ui.fragment
 * FileName: GankMeiZiDetailFragment
 * Date: on 2018/6/16  下午6:42
 * Auther: zihe
 * Descirbe:
 * Email: hezihao@linghit.com
 */
class GankMeiZiDetailFragment : BaseMvpFragment<GankMeiZiDetailContract.Presenter, GankMeiZiDetailContract.View>(), GankMeiZiDetailContract.View {
    private lateinit var mImageView: PhotoView
    private lateinit var mToolbar: Toolbar
    private lateinit var mLoadingView: RotateCircleProgressBar

    private var isHideToolBar = false
    private var mImageBean: DisplayMeiZiImageBean? = null

    override fun onDestroy() {
        showStatusBar()
        super.onDestroy()
    }

    override fun onLayoutId(): Int {
        return R.layout.fragment_gank_meizi_detail
    }

    override fun onFindView(rootView: View) {
        super.onFindView(rootView)
        mImageView = findView(R.id.image_iv)
        mLoadingView = findView(R.id.loading_iv)
    }

    override fun onBindViewContent() {
        super.onBindViewContent()
        val toolBarHelper = ToolBarHelper.newBuilder(view, R.id.tool_bar, object : ToolBarHelper.ConfigCallbackAdapter() {
            override fun onConfigBefore(toolbar: Toolbar) {
                (activity as AppCompatActivity).setSupportActionBar(toolbar)
            }
        })
                .withNavigationIconDrawable(R.drawable.ic_action_back)
                .withNavigationIconOnClick { activity!!.onBackPressed() }
                .build()
        mToolbar = toolBarHelper.toolbar
        ImmersionBar.with(this).titleBar(mToolbar!!)
        //长按点击事件
        mImageView.setOnPhotoTapListener { view, x, y -> toggleToolBar() }
    }

    private fun toggleToolBar() {
        //隐藏
        if (!isHideToolBar) {
            mToolbar.animate()
                    .translationY((-mToolbar.height).toFloat())
                    .setInterpolator(AccelerateInterpolator())
                    .start()
            hideStatusBar()
        } else {
            //显示
            mToolbar.animate()
                    .translationY(0f)
                    .setInterpolator(AccelerateInterpolator())
                    .start()
            showStatusBar()
        }
        isHideToolBar = !isHideToolBar
    }

    override fun setupSwipeBackEnable(): Boolean {
        return true
    }

    override fun onCreatePresenter(): GankMeiZiDetailContract.Presenter {
        return GankMeiZiDetailPresenter()
    }

    override fun onLayoutAfter() {
        super.onLayoutAfter()
        presenter.getMeiZiDetail(arguments!!)
    }

    override fun showMeiZiDetail(bean: DisplayMeiZiImageBean) {
        this.mImageBean = bean
        mLoadingView.visibility = View.VISIBLE
        ImageDisplayUtil.display(activity, bean.url, mImageView, object : ImageDisplayUtil.DisplayCallbackAdapter() {
            override fun onReady() {
                super.onReady()
                mLoadingView.visibility = View.GONE
            }
        })
    }

    override fun showError(throwable: Throwable) {

    }

    /**
     * 保存图片到相册
     */
    private fun saveImageToGallery() {
        mImageBean.let { it ->
            if (it?.url != null && it.createdAt != null) {
                val disposable = ImageDownloadUtil
                        .saveImageAndGetPathObservable(activity, mImageBean!!.url, mImageBean!!.createdAt)
                        .doOnSubscribe { showLoading("保存中，请稍后...") }
                        .doOnTerminate { hideLoading() }
                        .subscribe({
                            val appDir = File(Environment.getExternalStorageDirectory(), "Meizhi")
                            val msg = String.format(getString(R.string.picture_has_save_to),
                                    appDir.absolutePath)
                            EasyToast.newBuilder().build().show(msg)
                        }, { EasyToast.newBuilder().build().show("再试试...") })
                (presenter as BasePresenter<*>).addSubscription(disposable)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.let {
            it!!.inflate(R.menu.menu_detail, menu)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.action_share -> return true
            R.id.action_save -> {
                saveImageToGallery()
                return true
            }
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
